src/java/org/apache/cassandra/db/virtual/CollectionVirtualTableAdapter.java

```diff
@@ -0,0 +1,589 @@
+                    }
+                }
+            });
+
+        if (partitionKeyTypes.size() == 1)
+            builder.partitioner(new LocalPartitioner(partitionKeyTypes.get(0)));
+        else if (partitionKeyTypes.size() > 1)
+            builder.partitioner(new LocalPartitioner(CompositeType.getInstance(partitionKeyTypes)));
+
+        return builder.build();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public UnfilteredPartitionIterator select(DecoratedKey partitionKey,
+                                              ClusteringIndexFilter clusteringFilter,
+                                              ColumnFilter columnFilter)
+    {
+        if (!data.iterator().hasNext())
+            return EmptyIterators.unfilteredPartition(metadata);
+
+        NavigableMap<Clustering<?>, Row> rows = new TreeMap<>(metadata.comparator);
+        Stream<CollectionRow> stream;
+        if (decorateKeyToRowExtractor == null)
+        {
+            // The benchmark shows that if we continuously read the data from the virtual table e.g. by metric name,
+            // then the parallel stream is slightly faster to get the first result, but for continuous reads it gives us
+            // a higher GC pressure. The sequential stream is slightly slower to get the first result, but it has the
+            // same throughput as the parallel stream, and it gives us less GC pressure.
+            // See the details in the benchmark: https://gist.github.com/Mmuzaf/80c73b7f9441ff21f6d22efe5746541a
+            stream = StreamSupport.stream(data.spliterator(), false)
+                                  .map(row -> makeRow(row, columnFilter))
+                                  .filter(cr -> partitionKey.equals(cr.key.get()))
+                                  .filter(cr -> clusteringFilter.selects(cr.clustering));
+        }
+        else
+        {
+            R row = decorateKeyToRowExtractor.apply(partitionKey);
+            if (row == null)
+                return EmptyIterators.unfilteredPartition(metadata);
+            stream = Stream.of(makeRow(row, columnFilter));
+        }
+
+        // If there are no clustering columns, we've found a unique partition that matches the partition key,
+        // so we can stop the stream without looping through all the rows.
+        if (walker.count(Column.Type.CLUSTERING) == 0)
+            stream.findFirst().ifPresent(cr -> rows.put(cr.clustering, cr.rowSup.get()));
+        else
+            stream.forEach(cr -> rows.put(cr.clustering, cr.rowSup.get()));
+
+        return new SingletonUnfilteredPartitionIterator(new DataRowUnfilteredIterator(partitionKey,
+                                                                                      clusteringFilter,
+                                                                                      columnFilter,
+                                                                                      rows));
+    }
+
+    /** {@inheritDoc} */
```

Comment:
why do we have these inheritDoc here? e.g. IDEA just perfectly renders javadoc for this, taking it from the signature of method this overrides. I do not think this is necessary, there is a lot of instances like this in this patch

Reply:
Removed all such comments, I guess this is my old habit kicking in ;-)